core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -1148,10 +1150,26 @@ public ShardRouting next() {
                 if (hasNext() == false) {
                     throw new NoSuchElementException();
                 }
-                Iterator<ShardRouting> iter = queue.poll();
-                ShardRouting result = iter.next();
-                queue.offer(iter);
-                return result;
+                long startTime = System.currentTimeMillis();
+                while (hasNext()) {
+                    logger.info("Queue size {}", queue.size());
+                    Iterator<ShardRouting> iter = queue.poll();
+                    ShardRouting result = iter.next();
+                    RoutingNode routingNode = nodesToShards.get(result.currentNodeId());
+                    Decision decision = allocation.deciders().decideOutgoingMovePerNode(routingNode, allocation, routingNodes);
+                    if (decision == Decision.NO) {
+                        logger.info("Keeping node iterator " + routingNode.nodeId() + " " + decision.toString());
+                        queue.offer(iter);
+                        long endTime = System.currentTimeMillis();
+                        logger.info("Node level Allocation decision time {}", (endTime - startTime));
+                        return result;
+                    } else {
+                        logger.info("Removing node iterator " + routingNode.nodeId() + " " + decision.toString());
```

Comment:
Actually, the confusion is arising due to the method name- it should be renamed to canRemainOnNode instead of decideOutgoingMovePerNode. 
A throttle is a best decision which means that the node needs to be removed anyway, while YES means that we should iterate through all other deciders to take a decision. If all deciders return that the shards on this node can remain, we remove that node from the iterators.

Consider these scenarios( I think we should UTs for all of this.)
1. Node is not excluded, node disk is not full, forced allocation disabled and shard limit not breached and not enough outgoing concurrent recoveries on node - node should be removed from iteration.
2. Node is excluded, node disk is not full, forced allocation disabled and shard limit not breached and not enough outgoing concurrent recoveries on node - node should not be removed from iteration.
3. Node is excluded, node disk is not full, forced allocation disabled and shard limit not breached and maximum outgoing concurrent recoveries on node - node removed from iteration, as it cannot be moved anyway.
4. Node is not excluded, node disk is full, forced allocation disabled and shard limit not breached and maximum outgoing concurrent recoveries on node - node removed from iteration, as shards cannot be moved anyway.
5. Node is not excluded, node disk is full, forced allocation disabled and shard limit not breached and not enough outgoing concurrent recoveries on node - node should not be removed from iteration.
6. Node is not excluded, node disk is not full, forced allocation disabled and shard limit breached and not enough outgoing concurrent recoveries on node - node should not be removed from iteration.

and so on and so forth..